Pr/0003 ctel zzhc

pkchan opened the pull request:
zzhc library, as required by China Telecom. It is by default disabled, so no example is available.

To enable, you need the following Kconfig options:
- CONFIG_ZZHC=y
- CONFIG_HEAP_MEM_POOL_SIZE=4096

As required by CONFIG_SETTINGS, which is one of the direct dependencies:
- CONFIG_FLASH=y
- CONFIG_FLASH_PAGE_LAYOUT=y
- CONFIG_FLASH_MAP=y
- CONFIG_MPU_ALLOW_FLASH_WRITE=y
- CONFIG_NVS=y
- CONFIG_SETTINGS_NVS=y
- CONFIG_SETTINGS_NVS_SECTOR_COUNT=6

As required by CONFIG_BSD_LIBRARY, which is another direct dependency:
- CONFIG_NETWORKING=y
- CONFIG_NET_NATIVE=n
- CONFIG_NET_SOCKETS=y
- CONFIG_NET_SOCKETS_OFFLOAD=y

To allow for automatic registration to operator's network, you need the following:
- CONFIG_LTE_LINK_CONTROL=y
- CONFIG_LTE_AUTO_INIT_AND_CONNECT=y

Depends on PR #1337 by @hakonfam in order to compile and function.

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:175: WARNING:NEW_TYPEDEFS: do not add new typedefs
#175: FILE: lib/zzhc/zzhc.c:43:
+typedef enum state state_t;

-:197: WARNING:NEW_TYPEDEFS: do not add new typedefs
#197: FILE: lib/zzhc/zzhc.c:65:
+typedef struct zzhc zzhc_t;

-:216: WARNING:LINE_SPACING: Missing a blank line after declarations
#216: FILE: lib/zzhc/zzhc.c:84:
+	zzhc_t *ctx = (zzhc_t *)context;
+	if (ctx == NULL) {

-:220: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#220: FILE: lib/zzhc/zzhc.c:88:
+	if (0 == memcmp(response, status_cereg, sizeof(status_cereg))) {

-:222: WARNING:SUSPECT_CODE_INDENT: suspect code indent for conditional statements (16, 19)
#222: FILE: lib/zzhc/zzhc.c:90:
+		if (rc < 0) {
+		   return;

-:223: WARNING:TABSTOP: Statements should start on a tabstop
#223: FILE: lib/zzhc/zzhc.c:91:
+		   return;

-:243: WARNING:LINE_SPACING: Missing a blank line after declarations
#243: FILE: lib/zzhc/zzhc.c:111:
+		char *cptr = strchr(data, tok);
+		if (cptr == NULL) {

-:247: WARNING:SPACING: space prohibited before semicolon
#247: FILE: lib/zzhc/zzhc.c:115:
+	} while (true) ;

-:256: ERROR:SPACING: spaces required around that '=' (ctx:VxV)
#256: FILE: lib/zzhc/zzhc.c:124:
+	int i=0, rc;
 	     ^

-:295: ERROR:CODE_INDENT: code indent should use tabs where possible
#295: FILE: lib/zzhc/zzhc.c:163:
+^I        ctx->at_resp, chip_rev[i]);$

-:489: WARNING:LINE_SPACING: Missing a blank line after declarations
#489: FILE: lib/zzhc/zzhc.c:357:
+	int len = recv(ctx->sock_fd, rx_buff, RX_BUF_LEN, 0);
+	LOG_DBG("len=%d", len);

-:525: ERROR:CODE_INDENT: code indent should use tabs where possible
#525: FILE: lib/zzhc/zzhc.c:393:
+^I                 AT_RESPONSE_LEN) != 0) {$

-:539: ERROR:TRAILING_WHITESPACE: trailing whitespace
#539: FILE: lib/zzhc/zzhc.c:407:
+^I}^I   $

-:550: ERROR:SWITCH_CASE_INDENT_LEVEL: switch and case should be at the same indent
#550: FILE: lib/zzhc/zzhc.c:418:
+	switch (ctx->state) {
+		case STATE_INIT:
[...]
+		case STATE_WAIT_FOR_REG:
[...]
+		case STATE_START:
[...]
+		case STATE_HTTP_REQ_SENT:
[...]
+		case STATE_HTTP_RESP_RECVD:
[...]
+		case STATE_RETRY_TIMEOUT:
[...]
+		case STATE_END:
[...]
+		default:

-:563: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#563: FILE: lib/zzhc/zzhc.c:431:
+			ctx->state = (0 == check_iccid(ctx))

-:564: ERROR:CODE_INDENT: code indent should use tabs where possible
#564: FILE: lib/zzhc/zzhc.c:432:
+^I^I^I           ? STATE_END$

-:565: ERROR:CODE_INDENT: code indent should use tabs where possible
#565: FILE: lib/zzhc/zzhc.c:433:
+^I^I^I           : STATE_START;$

-:579: ERROR:SPACING: space prohibited before that close parenthesis ')'
#579: FILE: lib/zzhc/zzhc.c:447:
+				if (post_request_send(ctx) != 0 ) {

-:642: ERROR:CODE_INDENT: code indent should use tabs where possible
#642: FILE: lib/zzhc/zzhc.c:510:
+^I^I^I           ? STATE_START : STATE_WAIT_FOR_REG;$

-:733: ERROR:CODE_INDENT: code indent should use tabs where possible
#733: FILE: lib/zzhc/zzhc.c:601:
+^I                CONFIG_KERNEL_INIT_PRIORITY_DEVICE, &api);$

-:768: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#768: FILE: lib/zzhc/zzhc_port.c:28:
+    printf(__VA_ARGS__);\$

-:769: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#769: FILE: lib/zzhc/zzhc_port.c:29:
+    printf("\n");\$

-:824: ERROR:CODE_INDENT: code indent should use tabs where possible
#824: FILE: lib/zzhc/zzhc_port.c:84:
+^I                thread_stack, K_THREAD_STACK_SIZEOF(thread_stack),$

-:825: ERROR:CODE_INDENT: code indent should use tabs where possible
#825: FILE: lib/zzhc/zzhc_port.c:85:
+^I                start_routine, arg, NULL, NULL,$

-:826: ERROR:CODE_INDENT: code indent should use tabs where possible
#826: FILE: lib/zzhc/zzhc_port.c:86:
+^I                THREAD_PRIORITY, 0, K_NO_WAIT);$

-:834: WARNING:LINE_SPACING: Missing a blank line after declarations
#834: FILE: lib/zzhc/zzhc_port.c:94:
+	u32_t ficr_addr = NRF_FICR_S_BASE + FICR_REV_OFFSET;
+	return spm_request_read(i, ficr_addr, sizeof(u32_t));

-:904: ERROR:SPACING: space prohibited before that close parenthesis ')'
#904: FILE: lib/zzhc/zzhc_port.c:164:
+		size_t slen )

-:913: ERROR:SPACING: spaces required around that '?' (ctx:VxW)
#913: FILE: lib/zzhc/zzhc_port.c:173:
+	return (rc == 0)? i: 0;
 	                ^

-:913: ERROR:SPACING: spaces required around that ':' (ctx:VxW)
#913: FILE: lib/zzhc/zzhc_port.c:173:
+	return (rc == 0)? i: 0;
 	                   ^

-:1026: ERROR:POINTER_LOCATION: "foo * bar" should be "foo *bar"
#1026: FILE: lib/zzhc/zzhc_port.c:286:
+	cJSON * root_obj;

-:1027: ERROR:POINTER_LOCATION: "foo * bar" should be "foo *bar"
#1027: FILE: lib/zzhc/zzhc_port.c:287:
+	cJSON * resultCode_obj;

-:1028: ERROR:POINTER_LOCATION: "foo * bar" should be "foo *bar"
#1028: FILE: lib/zzhc/zzhc_port.c:288:
+	cJSON * resultDesc_obj;

-:1051: ERROR:SPACING: spaces required around that ':' (ctx:VxW)
#1051: FILE: lib/zzhc/zzhc_port.c:311:
+			? true: false;
 			      ^

-:1063: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#1063: FILE: lib/zzhc/zzhc_port.c:323:
+		if (0 == strcmp(resultDesc_obj->valuestring, "Success")) {

-:1073: ERROR:TRAILING_WHITESPACE: trailing whitespace
#1073: FILE: lib/zzhc/zzhc_port.c:333:
+ * $

- total: 21 errors, 14 warnings, 1064 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

NOTE: Whitespace errors detected.
      You may wish to use scripts/cleanpatch or scripts/cleanfile

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.
```

## Codeowners issues

```
New files added that are not covered in CODEOWNERS:

lib/zzhc/CMakeLists.txt
lib/zzhc/zzhc.c
lib/zzhc/zzhc_port.c

Please add one or more entries in the CODEOWNERS file to cover those files
```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

SebastianBoe commented on lib/zzhc/zzhc_port.c:
```diff
@@ -0,0 +1,364 @@
+static void this_free(void *ptr)
+{
+	k_free(ptr);
+}
```
I don't see the value in these wrappers. k_free should be invoked directly instead of indirectly.

SebastianBoe replied:
The same applies for other wrappers here.

pkchan replied:
zzhc.c and zzhc_port.c will be merged as one file right before the end of this review, in this PR.
Then you will not see any indirect calls like this.
In the next few days they still need to be kept. Sorry about this.

SebastianBoe commented on lib/Kconfig:
```diff
@@ -22,4 +22,6 @@ rsource "modem_info/Kconfig"
 rsource "pdn_management/Kconfig"
 
+rsource "zzhc/Kconfig"
+
```
The two commits should be one.

It doesn't make sense to include one without the other.

pkchan replied:
Will squash them into single commit.

pkchan replied:
Squashed.

SebastianBoe commented on lib/zzhc/Kconfig:
```diff
@@ -0,0 +1,41 @@
+	select FLASH_PAGE_LAYOUT
+	select FLASH_MAP
+	select MPU_ALLOW_FLASH_WRITE
+	select SETTINGS
```
We need to hide this option from unrelated platforms. E.g.

depends on BSD_LIB

or similar.

pkchan replied:
The original idea was to have only one (or a few) options to enable the module.

As you can see, having 12 options in order to enable only one library is prone to error, and can be frustrating to users of the library, including customers.

How about this way?
- keeps selecting AT_CMD, AT_CMD_PARSER, AT_NOTIF, BASE64, CJSON_LIB, NEWLIB_LIBC
- depends on SETTINGS
- put FLASH, FLASH_PAGE_LAYOUT, FLASH_MAP, MPU_ALLOW_FLASH_WRITE, NVS and SETTINGS_NVS into prj.conf. (This is learned from example\bluetooth\peripheral_hids_keyboard.)

For users of the SDK, simple is the best.

SebastianBoe replied:
I tried the new version. Using the hello_world sample and the nrf9160_pca10090ns board I found that after first enabling SETTINGS, and then ZZHC, i still ended up with a corrupted configuration because AT_CMD depends on BSD_LIBRARY and could therefore not be selected by ZZHC.

What if we change

depends on SETTINGS

to

depends on BSD_LIBRARY
select SETTINGS

then we still only have to enable one option, but we don't get a corrupted configuration. or

depends on TRUSTED_EXECUTION_NONSECURE
select SETTINGS
select BSD_LIBRARY

if you would prefer to not have the user need to enable anything.

pkchan replied:
@SebastianBoe The same applies to nrf_cloud libraries as well. If you need to make it work in hello_world, then you need to have lots of options in prj.conf to work with (because Kconfig does not support deselection and value assignment).

I have changed the Kconfig file so that all direct dependencies are selected:
`depends on TRUSTED_EXECUTION_NONSECURE`
`select BSD_LIBRARY`
`select SETTINGS`
`select AT_CMD`
`select AT_CMD_PARSER`
`select AT_NOTIF`
`select BASE64`
`select CJSON_LIB`
`select NEWLIB_LIBC`

To make it work on zephyr\samples\hello_world, you need the followings in prj.conf:
`CONFIG_ZZHC=y`

`# Required by Settings`
`CONFIG_FLASH=y`
`CONFIG_FLASH_PAGE_LAYOUT=y`
`CONFIG_FLASH_MAP=y`
`CONFIG_MPU_ALLOW_FLASH_WRITE=y`
`CONFIG_NVS=y`
`CONFIG_SETTINGS_NVS_SECTOR_COUNT=6`

`# Requied by k_malloc()`
`CONFIG_HEAP_MEM_POOL_SIZE=4096`

`# Network`
`CONFIG_NETWORKING=y`
`CONFIG_NET_NATIVE=n`
`CONFIG_NET_SOCKETS=y`
`CONFIG_NET_SOCKETS_OFFLOAD=y`

`# LTE link control`
`CONFIG_LTE_LINK_CONTROL=y`
`CONFIG_LTE_AUTO_INIT_AND_CONNECT=y`

SebastianBoe commented on lib/zzhc/zzhc.c:
```diff
@@ -0,0 +1,600 @@
+		}
+		*cptr = substitute;
+	} while (true);
+}
```
This is O(N*N).

Let's do a single pass ( O(N) ).

for(int i = 0; i < strlen(data); i++)
{ if (data[i] == tok) data[i] = substitue; }

pkchan replied:
I prefer calling library as much as possible:
`	char *cptr = data;

	while (cptr != NULL) {
		if ((cptr = strchr(cptr, tok)) != NULL) {
			*cptr++ = substitute;
		}
	}`

SebastianBoe replied:
I find the version without strchr easier to read (and probably faster assuming strchr is not inlined), but at least it's now O(N), so if you prefer using strchr then that is OK.

pkchan replied:
The solution having strlen() is 2-pass. data is traversed two times: the first time searching for NULL, and the second time searching for tok.

(Disclaimer: I'm not expert in algorithms...... )

SebastianBoe replied:
My mistake, I thought strlen was O(1).

using the revised strchr version is fine.

pkchan commented:
@rlubos I've found that this zzhc library should be moved from nrf\lib\zzhc to nrf\subsys\net\lib\zzhc, as there are something similar inside. e.g. nrf_cloud, download_client...etc. How do you think?

rlubos commented:
> I've found that this zzhc library should be moved from nrf\lib\zzhc to nrf\subsys\net\lib\zzhc

Sounds reasonable, it is a networking library after all.

rlubos commented on lib/zzhc/zzhc.c:
```diff
@@ -0,0 +1,600 @@
+	}
+
+	/* Allocate resource. */
+	ctx = (struct zzhc *)this_malloc(sizeof(struct zzhc));
```
Could we use a static memory instead of heap for context data? Zephyr heap is known for not-handling-well large allocations, so this might cause trouble for smaller heaps. It is under a rework, but who knows when will this be fixed...

pkchan replied:
This is the tricky part.

The zzhc specification says, "After execution, it should quit and not be persistent." Therefore, the memory need to be freed. With `k_free()`, we need `k_malloc()` 

Besides, using this kind of malloc-and-free approach (i.e. new and delete object) allows for higher flexibility in the future so that multiple instances of the library can be instantiated. With slight modifications in the future, we can launch multiple threads to connect to different servers.

The allocated memory fragments are not really large:
- struct zzhc has 644 bytes. (This will become 640 bytes in the next push.)
- struct k_thread has 176 bytes. (This will become static in the next push.)
- struct k_sem has 24 bytes.
- struct at_param_list has 8 bytes.
- struct settings_handler has 24 bytes.

Will these make Zephyr fail?

rlubos replied:
Not exactly fail, but even all these allocations are < 1 kB, they would require ~3kB heap size with current Zephyr heap implementation.

pkchan replied:
`thread `has been removed, whereas `storage `has become static variable global to the file.
While the zzhc_port.c is still needed at the moment, `sem` and `at_list` still need to be allocated. They will become part of `struct zzhc` once zzhc_port.c is merged to zzhc.c.

rlubos commented on lib/zzhc/zzhc_port.c:
```diff
@@ -0,0 +1,365 @@
+static char const settings_name[] = "zzhc";
+
+#ifndef LOG_DBG
+#define LOG_DBG(...) \
```
Why is this needed, doesn't the logger provide a macro on its own?

pkchan replied:
Some 3rd parties are using this. As mentioned earlier, zzhc.c and zzhc_port.c will be merged into one file near the end of the review of this PR. This dummy stuff will be removed then.

pkchan replied:
In the push on 5th Nov, 2019, the calls to printf() inside such macro have been removed. As said, the library is still being used by 3rd parties, there is a need to leave it temporarily here. As also said, once zzhc.c and zzhc_port.c are merged, such macro will be eliminated.

rlubos commented on lib/zzhc/zzhc_port.c:
```diff
@@ -0,0 +1,365 @@
+		return -ENOBUFS;
+	}
+
+	ctx->sem = this_malloc(sizeof(struct k_sem));
```
Do w really need to use heap for all these allocations?

pkchan replied:
Perhaps we can continue discussion [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#discussion_r339536848).

rlubos commented on lib/zzhc/zzhc_port.c:
```diff
@@ -0,0 +1,365 @@
+	cJSON *resultCode_obj;
+	cJSON *resultDesc_obj;
+
+	cJSON_Init();
```
Would it be possible to use Zephyr json library instead? https://github.com/zephyrproject-rtos/zephyr/blob/master/include/data/json.h

The cJSON library was brought into NCS as a dependency for nRF Cloud library, and the long-term plan is to drop its usage in favour of Zephyr json lib.

pkchan replied:
Good idea. (since there is example to learn in NCS)
I'll try to re-implement check_http_payload() using Zephyr json library in the next push.

hakonfam replied:
FYI `cJSON` will be used by `aws` libraries. The reason this is done instead of zephyr json is that the zephyr json does not support floating point (I have been told), and floats are required by the thingy project (which makes sense). Maybe we should take this discussion elsewhere, but thought you should know.

rlubos replied:
Ok, I was not aware, so it could be that my comment is not a valid one.

Still, in my opinion, it makes little sense to maintain an external library, if we have a native counterpart in Zephyr. If it is missing a feature, it could be extended.

pkchan replied:
In the last push on 4th Nov 2019, cJSON has been replaced by zephyr's json library.

pkchan commented on lib/zzhc/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+#
+# Copyright (c) 2018 Nordic Semiconductor
```
Will change from 2018 to 2019 in the next commit.

pkchan replied:
Done in the last push on 4th Nov 2019.

pkchan commented on lib/zzhc/Kconfig:
```diff
@@ -0,0 +1,39 @@
+	select AT_NOTIF
+	select BASE64
+	select CJSON_LIB
+	select NEWLIB_LIBC
```
NEWLIB_LIBC is required for cJSON to be compiled.

pkchan replied:
In the last push on 4th Nov 2019, cJSON library has been replaced by zephyr's json library. Therefore, NEWLIB_LIBC is not needed anymore.

pkchan commented on lib/zzhc/zzhc.c:
```diff
@@ -0,0 +1,600 @@
+static struct api const api = {NULL};
+
+/**@brief Device definition for zzhc */
+DEVICE_AND_API_INIT(zzhc, "ZZHC", zzhc_init, NULL, NULL, APPLICATION,
```
Yes, it could be done.

The original design of at_cmd notification does not allow for passing of context pointers to receiving module. The intention of using DEVICE_AND_API_INIT() is to make sure the at-notification handler can get the pointer to context through the use of device_get_binding(). However, now the new design in at_notif allows for passing of pointer to context. Therefore, we do not need DEVICE_AND_API_INIT().

There is another consideration which I do not know if the K_THREAD_DEFINE() approach supports or not. It is that zzhc_init() has to happen after at_cmd, at_notif, nvs and settings are ready. DEVICE_AND_API_INIT() and SYS_INIT() allows for such. Can the K_THREAD_DEFINE()() also support so?

rlubos replied:
Yeah, I see your point... As for the `K_THREAD_DEFINE()`, all threads are fired up just before the main() starts, so after all initializations are complete, so there should be no harm in initializing within a thread itself. 

But that's how it works, I haven't found any formal confirmation of this behavior in the docs. `SYS_INIT()` sounds fine for library initialization as well, I forgot about this function. So I leave it up to you whether to use `SYS_INIT()`, or initialize in the thread directly.

pkchan replied:
I vote for K_THREAD_DEFINE(), because it is simpler than SYS_INIT().
We also need to care about the startup sequence of SYS_INIT() any more.

pkchan replied:
Done in the last push on 4th Nov 2019.

pkchan commented on lib/zzhc/zzhc.c:
```diff
@@ -0,0 +1,600 @@
+#include <logging/log.h>        /**< Required by logging. */
+#define LOG_DOMAIN zzhc
+#define LOG_LEVEL CONFIG_ZZHC_LOG_LEVEL
+LOG_MODULE_REGISTER(LOG_DOMAIN);
```
OK, will change.

pkchan replied:
Done in the last push on 4th Nov 2019.

pkchan commented on lib/zzhc/zzhc.c:
```diff
@@ -0,0 +1,600 @@
+#define JSON_MAX_LEN      320       /** Length of json */
+#define IMEI_LEN          15        /** Length of IMEI */
+#define ICCID_LEN         20        /** Length of ICCID */
+#define RETRY_TIMEOUT     (3600)      /** 1 hour, 3600s */
```
OK, will change in the next push.

pkchan replied:
Done in the last push on 4th Nov 2019.

pkchan commented on lib/zzhc/zzhc.c:
```diff
@@ -0,0 +1,600 @@
+
+	if (rc != 0) {
+		LOG_DBG("Unable to connect.");
+		close(ctx->sock_fd);
```
close() does nothing on invalid socket handle. It does not matter whether `ctx->sock_fd` is valid or not.

rlubos replied:
I'd argue if returning an error is nothing, but ok, I get your point.

pkchan commented on lib/zzhc/zzhc.c:
```diff
@@ -0,0 +1,600 @@
+/**@brief +CEREG event prefix. */
+static const char status_cereg[7] = "+CEREG:";
+
+/**@brief Lookup table for chip revision. */
```
It's a 4-byte LUT:
0 --> X
1 --> A
2 --> 0
3 --> 1

AND, you reminded me of one thing: array index out of bound when reading `chip_rev[i]`. For example, `i = 4`. I will add some protection to it in the next push.

```
	if (i >= sizeof(chip_rev)) {
		i = sizeof(chip_rev) - 1;
	}
```

pkchan replied:
Added protection mechanism in the last push on 4th Nov 2019.

pkchan commented on lib/zzhc/zzhc.c:
```diff
@@ -0,0 +1,600 @@
+/**@brief Lookup table for chip revision. */
+static const char chip_rev[4] = "XA01";
+
+#include "zzhc_port.c"
```
Yes, please kindly ignore this right now. There is still a 3rd party who need a separation between zzhc and zzhc_port.

As mentioned earlier, the two files will be merged into one near the end of this PR review.

pkchan replied:
In the 2nd push on 5th Nov, 2019 (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#event-2770493390)), the inclusion was eliminated.

pkchan commented on lib/zzhc/zzhc.c:
```diff
@@ -0,0 +1,600 @@
+		}
+
+		LOG_DBG("+CEREG: %d", rc);
+		if (rc == 1 || rc == 5) {
```
OK, will change in the next push.

pkchan replied:
Done in the last push on 4th Nov 2019.

pkchan commented on lib/zzhc/zzhc.c:
```diff
@@ -0,0 +1,600 @@
+	strcat(ctx->http_pkt, "Connection: close\r\n");
+	strcat(ctx->http_pkt, "Content-Length: ");
+
+	rc = build_json(ctx, json_buff, JSON_MAX_LEN);
```
In the push on 5th Nov, 2019, the return code from `build_json()` is propagated to `build_http_req()`, and then to `fsm()`. In case of errors, it will result in a state change to `STATE_RETRY_TIMEOUT` for further retries.

pkchan commented:
What has been changed since 28th Oct, 2019:
- Eliminated the inclusion of `zzhc_port.c` from `zzhc.c` (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#discussion_r339512801)).
- Added `lib\zzhc\zzhc_internal.h`, a place-holder for platform-dependent implementations, with details of APIs inside.
- Changed the "year" in file header from 2018 to 2019 (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#discussion_r339505365)).
- Replaced `cJSON` library from Zephyr's `json` library (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#discussion_r339508508)).
- Replaced `DEVICE_AND_API_INIT()` by `K_THREAD_DEFINE()` (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#discussion_r339510412)).
- Used single-line `LOG_MODULE_REGISTER()` (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#discussion_r339510687)).
- Removed unnecessary brackets (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#discussion_r339511672)).
- Added protection mechanism for LUT index (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#discussion_r339627772)).
- Used defines from `lte_lc` (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#discussion_r339514751)).
- Handled the return code from `build_json()` (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#discussion_r339519722)).
- Eliminated the dummy implementation in the `LOG_DBG()` inside `zzhc_port.c` (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#discussion_r339537930)).
- Made indirect calls to kernel functions to macros, including `k_sem_init()`,
  `k_sem_take()`, `k_sem_give()`, `k_sleep()`, `k_malloc()`, `k_free()`  (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#discussion_r339019165)).
- Renamed functions / macros, in order to avoid name-space conflict with other files, since functions in `zzhc_port.c` are not static anymore:
    - From `sem_init()` to `zzhc_sem_init()`
    - From `sem_wait()` to `zzhc_sem_wait()`
    - From `sem_post()` to `zzhc_sem_post()`
    - From `sleep()` to `zzhc_sleep()`
    - From `this_malloc()` to `zzhc_malloc()`
    - From `this_free()` to `zzhc_free()`
    - From `read_chip_rev()` to `zzhc_read_chip_rev()`
    - From `save_iccid()` to `zzhc_save_iccid()`
    - From `load_iccid()` to `zzhc_load_iccid()`
    - From `at_cmd_xfer()` to `zzhc_at_cmd_xfer()`
    - From `register_handler()` to `zzhc_register_handler()`
    - From `deregister_handler()` to `zzhc_deregister_handler()`
    - From `base64()` to `zzhc_base64()`
    - From `ext_init()` to `zzhc_ext_init()`
    - From `ext_uninit()` to `zzhc_ext_uninit()`
    - From `check_http_payload()` to `zzhc_check_http_payload()`
    - From `get_cereg_evt()` to `zzhc_get_at_param_short()`
- `zzhc_get_at_param_short()` to support select-able parameters by index.
- Documentation in `include\zzhc.rst`

Will do after the next review:
- Move from `nrf\lib\zzhc` to `nrf\subsys\net\lib\zzhc`  (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#issuecomment-546806862))

pkchan commented:
In the push on 6th Nov, 2019, `check_simcard_whitelist()` was added to make sure only China Telecom simcards are allowed to trigger the state machine.

Whitelisted PLMNs include:
- mcc=460, mnc=03, China Telecom
- mcc=460, mnc=05, China Telecom
- mcc=460, mnc=11, China Telecom
- mcc=455, mnc=02, China Telecom (Macau)

Note: MCC and MNC can be extracted from the 1st 5~6 bytes of IMSI.

rlubos commented on lib/zzhc/zzhc_port.c:
```diff
@@ -0,0 +1,191 @@
+
+/**@brief Structure to store decoded json object. */
+struct rsp_obj {
+	const char *resultCode;
```
Why camel case all of sudden, can we stick to the snake case with variables/struct members?

pkchan replied:
In the adoption to Zephyr's json library, I have used `JSON_OBJ_DESCR_PRIM()`, which requires struct member names to be identical to the json name fields. To use snake case, `JSON_OBJ_DESCR_PRIM_NAMED()` has to be used.

In the next push, the struct member names will become snake case, and `JSON_OBJ_DESCR_PRIM_NAMED()` will be used instead.

pkchan replied:
Done in the last push on 7th Nov, 2019 (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#event-2777928390)).

rlubos commented on lib/zzhc/zzhc_port.c:
```diff
@@ -0,0 +1,191 @@
+bool zzhc_check_http_payload(struct zzhc *ctx)
+{
+	int rc;
+	bool resultCode, resultDesc = false;
```
Same here, snake_case please.

pkchan replied:
To be discussed [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#discussion_r343123823).

rlubos commented on lib/zzhc/zzhc.c:
```diff
@@ -0,0 +1,631 @@
+
+	strcat(ctx->http_pkt, "User-Agent: zzhc/1.0.0\r\n");
+	strcat(ctx->http_pkt, "Accept: */*\r\n");
+	strcat(ctx->http_pkt, "x-forwarded-for: 117.61.5.70\r\n");
```
What's the IP address, is it ok to have it hardcoded in the request?

pkchan replied:
It's a constant according to the specification, so it should not matter to be hard-coded.

rlubos commented on lib/zzhc/zzhc.c:
```diff
@@ -0,0 +1,631 @@
+
+		LOG_DBG("+CEREG: %d", rc);
+		if (rc == LTE_LC_NW_REG_REGISTERED_HOME
+			|| rc == LTE_LC_NW_REG_REGISTERED_ROAMING) {
```
nit: Could possibly keep the `||` operator in the line above, for consistency.

pkchan replied:
Will be changed accordingly in the next push.

pkchan replied:
Done in the last push on 7th Nov, 2019 (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#event-2777928390)).

rlubos commented:
@ru-fu @b-gent Please review docs at convenience (@pkchan this might get delayed due to release though).

greg-fer commented on include/net/zzhc.rst:
```diff
@@ -0,0 +1,68 @@
+Telecommunication modules sold in China Telecom network requires registration of certain information on the module to the server (zzhc.vnet.cn).
+The ZZHC library implements chapter 6 of the "China Telecom IoT Module Requirements Whitepaper".
+As a background thread launched on boot, it automatically detects the availability of operator network.
+Once internet connectivity is ready and there is a new simcard which is inserted on boot, it proactively uploads IMEI, model number (i.e. nRF9160), modem revision, and the ICCID and IMSI of the simcard connected to nRF9160.
```
There are more than two things that are sent, so I would suggest using a bullet point list:
```suggestion
Once the Internet connectivity is ready and a new SIM card is detected on boot, the library proactively uploads the following information to the ZZHC server:
- IMEI
- model number (that is, nRF9160)
- modem revision
- nRF9160 SIM ICCID
- nRF9160 SIM IMSI
```

pkchan commented on include/net/zzhc.rst:
```diff
@@ -0,0 +1,68 @@
+The ZZHC library implements chapter 6 of the "China Telecom IoT Module Requirements Whitepaper".
+As a background thread launched on boot, it automatically detects the availability of operator network.
+Once internet connectivity is ready and there is a new simcard which is inserted on boot, it proactively uploads IMEI, model number (i.e. nRF9160), modem revision, and the ICCID and IMSI of the simcard connected to nRF9160.
+On any failure, which include unknown/negative response from server or any error during program execution, the ZZHC library retries the next hour for at most 5 times.
```
Per your questions:
1. The library retries sending the data to the server, correct?
Ans: Yes

2. Does the ZZHC library retries the next hour for at most 5 times mean the library retries 5 times each hour or once per hour, for a maximum 5 attempts? If the first option is correct, when does this process stop?
Ans: "once per hour, for a maximum 5 attempts"

greg-fer replied:
Thank you!

pkchan commented on include/net/zzhc.rst:
```diff
@@ -0,0 +1,68 @@
+
+Telecommunication modules sold in China Telecom network requires registration of certain information on the module to the server (zzhc.vnet.cn).
+The ZZHC library implements chapter 6 of the "China Telecom IoT Module Requirements Whitepaper".
+As a background thread launched on boot, it automatically detects the availability of operator network.
```
In the latest push, there is a new line right after line 12.

pkchan commented:
In the last push on 13th Nov, 2019,
- Applied changes per @greg-fer 's comments (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#pullrequestreview-315535630))
- Moved codes related to `lte_lc.h`, which is platform dependent, to `zzhc_internal.h`.
- Squashed commits.

greg-fer commented on include/net/zzhc.rst:
```diff
@@ -0,0 +1,74 @@
+To enable the library so that the end product can be approved by China Telecom, enable Kconfig option CONFIG_ZZHC.
+In prj.conf, set CONFIG_ZZHC=y.
+
+The thread behavior can also be configured using Kconfig options:
```
Do I understand correctly these two thread behavior options are not mandatory?

pkchan replied:
No, priority and stack size are supposed to be tweaked by SDK users.

greg-fer replied:
OK, no further edits needed.

greg-fer commented on include/net/zzhc.rst:
```diff
@@ -0,0 +1,74 @@
+   * To configure the stack size for the thread, change CONFIG_ZZHC_STACK_SIZE.
+   * To configure the thread priority, change CONFIG_ZZHC_THREAD_PRIO.
+
+The minimal heap size required is 4096 bytes.
```
Does this heap size requirement refer to the thread behavior options in line 49-50 or to enabling the library in general?

pkchan replied:
No, heap size is irrelevant to thread behavior.

Thread configurations govern how the thread executes. As a naive example, execute 50 lines, then sleep 10ms, then continue execution...etc.

Heap size governs how the library allocate and free memory. If the size is too small, the thread simply kills itself because of memory fault. If it is too large, then it will be a waste of memory.

pkchan replied:
Besides, stack size is not supposed to be touched unless the SDK user need to optimize and push to limits.

Heap size need to be accumulated to the heap size requirements of user application. For example, if the user application requires 16KB heap to run, then to enable zzhc library the SDK user need to increase it to 20KB.

greg-fer commented on include/net/zzhc.rst:
```diff
@@ -0,0 +1,74 @@
+
+.. _lib_zzhc_dependencies:
+
+Dependencies
```
Given that most of the library documentation does not feature a Dependencies section, I've moved this under Configuration.

pkchan replied:
Yes, thanks.

greg-fer commented on include/net/zzhc.rst:
```diff
@@ -0,0 +1,74 @@
+To allow for automatic registration to Lte-M or NB-IoT network on boot, the following are required in prj.conf:
+   * CONFIG_LTE_LINK_CONTROL=y
+   * CONFIG_LTE_AUTO_INIT_AND_CONNECT=y
+
```
Just checking: Is there going to be any API documentation appended to this document? All other libraries have the API documentation section included in their respective rst files.

pkchan replied:
No. This zzhc library does not have public API, so there is no documentation.

Instead, it's just a background thread which sends critical information to cloud quietly.

greg-fer commented on include/net/zzhc.rst:
```diff
@@ -0,0 +1,74 @@
+.. _lib_zzhc:
```
This library is included on the list of "Libraries for FOTA updates" on the Libraries section main page.
How does this library relate to the FOTA update libraries listed there? Could you clarify this in the documentation, @pkchan ?

pkchan replied:
`zzhc.rst` has nothing to do with FOTA. The file is in the `include\net` directory, which means it's the document for network libraries. I do not know why `include\net` has the name "**Libraries for FOTA updates**". It should be changed to "**Libraries for networking**" instead.

pkchan replied:
Per request, in `doc\nrf`, added section "**Libraries for networking**", which includes `zzhc`.

pkchan commented:
@greg-fer you reminded me of one thing: I need to find out the optimized `CONFIG_HEAP_MEM_POOL_SIZE` after changes required by reviewers of this PR. I will do so and update `zzhc.rst` soon.

pkchan commented:
> @greg-fer you reminded me of one thing: I need to find out the optimized `CONFIG_HEAP_MEM_POOL_SIZE` after changes required by reviewers of this PR. I will do so and update `zzhc.rst` soon.

In zzhc.rst, updated the requirement on `CONFIG_HEAP_MEM_POOL_SIZE` from 4096 to 2560.

greg-fer commented on doc/nrf/libraries.rst:
```diff
@@ -32,6 +32,11 @@ Here you can find documentation for these libraries, including API documentation
+   :maxdepth: 1
+   :caption: Libraries for networking:
+   :glob:
+   ../../include/net/zzhc/*
```
Please add a blank line before this line for this section to appear on the Libraries page.

pkchan replied:
Added and pushed.

rlubos commented:
FYI, the reason for the current doc failure:
```
/jenkins_workspace/workspace/connect-nrf_documentation_master@2/_build/nrf/rst/doc/nrf/include/net/zzhc/zzhc.rst:16: WARNING: Unexpected indentation.
```
@greg-fer Any hints on how to fix this?

greg-fer commented on include/net/zzhc/zzhc.rst:
```diff
@@ -0,0 +1,86 @@
+
+The ZZHC library is started as a background thread on boot and it automatically detects the availability of the operator network.
+Once the Internet connectivity is ready and a new SIM card is detected on boot, the library proactively uploads the following information to the ZZHC server:
+    * IMEI
```
To fix the indentation error, add a blank line before this line.

rlubos commented:
Some changes are still expected, added DNM so it's not merged prematurely.

pkchan commented:
What happened in the push on 21st Nov, 2019:

**Bug fix:**
- In function `zzhc_get_at_param_short()`, the check of the return value of `at_params_valid_count_get()` was wrong (`rc < 2`). It's corrected to (`rc < idx + 2`).

**Further changes** (as a new commit, to be squash after review by consultants):
- Used `%XOPERID` to filter simcards.
- Used `%SHORTSWVER` instead of `+CGMR` to get software version.
- Wakeup FSM thread from sleeping during state `STATE_RETRY_TIMEOUT` on reception of `%XSIM: 1` notification.
- ICCID became 19 bytes, and its byte-order was swapped.

pkchan commented:
Got green light from consultants.
I have squashed changes and then rebased to master. However, the last known good master is 4b5a6255. After that, there is an issue when initializing NVS. I still force-pushed such rebase to check the compliance to CI-system.

@rlubos can you advice how to go ahead on this issue?

About the issue:
- During startup, the NVS initialize. `get_flash_area_from_id(6)` returned -22. Actually,` flash_map` only consists of 3 elements, therefore returned error.

rlubos commented:
Hi @pkchan,
It seems that some changes were merged into the NVS settings subsystem during recent upmerge, which are not "partition manager aware" (https://github.com/NordicPlayground/fw-nrfconnect-zephyr/blob/master/subsys/settings/src/settings_init.c#L155). It should be easy to fix, I'll post a PR.

rlubos commented:
@pkchan The fix for NVS settings is now available on the master branch, please rebase and let me know if it's all good now.

pkchan commented:
> @pkchan The fix for NVS settings is now available on the master branch, please rebase and let me know if it's all good now.

@rlubos Rebased to daily master (31936f7b, 5am 26th Nov, 2019 HKT). It is working fine, and has been force-pushed.

metc commented on subsys/net/lib/zzhc/zzhc.c:
```diff
@@ -0,0 +1,639 @@
+
+		/* Clean-up */
+		disconnect(ctx);
+		zzhc_deregister_handler(ctx, at_notif_handler);
```
Should you not deregister in any case when the state machine is finished in case an error occurs in a previous state?

pkchan replied:
For any error happening in the state machine, the state `STATE_RETRY_TIMEOUT` will be reached. After `RETRY_CNT_MAX` (i.e. 3) times of any error, `STATE_END` will be reached. Such state refers to the end of execution.

The document from China Telecom states that "After execution, it should quit and not be persistent." This implies that `ctx` should be freed and `at_notif_handler` de-registered.
